fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java

```diff
@@ -0,0 +1,329 @@
+                    "partition " + pbi.partitionId + "-" + pbi.indexId);
+
+            Long minReplicaCount = pbi.beByReplicaCount.keySet().first();
+            Long maxReplicaCount = pbi.beByReplicaCount.keySet().last();
+            LOG.info("balancing partition {}-{} with replica count skew {} (min_replica_count: {}, max_replica_count: {})",
+                    pbi.partitionId, pbi.indexId, maxPartitionSkew,
+                    minReplicaCount, maxReplicaCount);
+
+            // Compute the intersection of the bes most loaded for the table
+            // with the bes most loaded overall, and likewise for least loaded.
+            // These are our ideal candidates for moving from and to, respectively.
+            IntersectionResult maxLoaded = getIntersection(ExtremumType.MAX, pbi.beByReplicaCount, beByTotalReplicaCount);
+            IntersectionResult minLoaded = getIntersection(ExtremumType.MIN, pbi.beByReplicaCount, beByTotalReplicaCount);
+            LOG.info("partition-wise: min_count: {}, max_count: {}", minLoaded.replicaCountPartition, maxLoaded.replicaCountPartition);
+            LOG.info("cluster-wise: min_count: {}, max_count: {}", minLoaded.replicaCountTotal, maxLoaded.replicaCountTotal);
+            LOG.debug("min_loaded_intersection: {}, max_loaded_intersection: {}", minLoaded.intersection.toString(), maxLoaded.intersection.toString());
+
+            // Do not move replicas of a balanced table if the least (most) loaded
+            // servers overall do not intersect the servers hosting the least (most)
+            // replicas of the table. Moving a replica in that case might keep the
+            // cluster skew the same or make it worse while keeping the table balanced.
+            if ((maxLoaded.replicaCountPartition <= minLoaded.replicaCountPartition + 1)
+                    && (minLoaded.intersection.isEmpty() || maxLoaded.intersection.isEmpty())) {
+                continue;
+            }
+
+            Long minLoadedBe, maxLoadedBe;
+            if (equalSkewOption == EqualSkewOption.PICK_FIRST) {
+                // beWithExtremumCount lists & intersection lists are natural ordering
+                minLoadedBe = minLoaded.intersection.isEmpty() ? minLoaded.beWithExtremumCount.get(0) : minLoaded.intersection.get(0);
+                maxLoadedBe = maxLoaded.intersection.isEmpty() ? maxLoaded.beWithExtremumCount.get(0) : maxLoaded.intersection.get(0);
+            } else {
+                minLoadedBe = minLoaded.intersection.isEmpty() ? getRandomListElement(minLoaded.beWithExtremumCount)
+                        : getRandomListElement(minLoaded.intersection);
+                maxLoadedBe = maxLoaded.intersection.isEmpty() ? getRandomListElement(maxLoaded.beWithExtremumCount)
+                        : getRandomListElement(maxLoaded.intersection);
+            }
+
+            LOG.debug("min_loaded_be: {}, max_loaded_be: {}", minLoadedBe, maxLoadedBe);
+            if (minLoadedBe.equals(maxLoadedBe)) {
```

Comment:
If we use PICK_FIRST, in the case of fewer candidate sets, will there be a high probability of the same?

Reply:
Now, we only use PICK_FIRST in UTs, to get the results that can be expected.